src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs

```diff
@@ -41,13 +41,25 @@ public override Task MatchAsync(HttpContext httpContext)
         private Matcher CreateMatcher(IReadOnlyList<Endpoint> endpoints)
         {
             var builder = _matcherBuilderFactory();
+            var seenEndpointNames = new List<string>();
             for (var i = 0; i < endpoints.Count; i++)
             {
                 // By design we only look at RouteEndpoint here. It's possible to
                 // register other endpoint types, which are non-routable, and it's
                 // ok that we won't route to them.
                 if (endpoints[i] is RouteEndpoint endpoint && endpoint.Metadata.GetMetadata<ISuppressMatchingMetadata>()?.SuppressMatching != true)
                 {
+                    // Validate that endpoint names are unique.
+                    var endpointName = endpoint.Metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName;
+                    if (endpointName != null)
+                    {
+                        if (seenEndpointNames.Contains(endpointName))
+                        {
+                            throw new Exception($"Duplicate endpoint name {endpointName} found. Endpoint names must be globally unique.");
```

Comment:
We never throw the base Exception, maybe InvalidOperationException.

Reply:
Yeah, `InvalidOperationException` didn't seem quite right either because it's moreso invalid data. I was trying to find an exception for that...

Reply:
Actually, maybe `InvalidOperationException` is fine...